src/main/java/com/google/research/bleth/services/DatabaseService.java

```diff
@@ -0,0 +1,225 @@
+        if (!isRoundExistsInSimulation(simulationId, round)) {
+            return null;
+        }
+        return toJsonTable(pq.asIterable(), simulationId);
+    }
+
+    private String toJsonTable(Iterable<Entity> boardStateEntities, String simulationId) {
+        // Retrieve simulations' dimensions.
+        Entity simulationEntity = retrieveSimulationEntity(simulationId);
+        int rowNum = ((Long) simulationEntity.getProperty("rowNum")).intValue();
+        int colNum = ((Long) simulationEntity.getProperty("colNum")).intValue();
+
+        // Create an empty table representing the board state.
+        ArrayTable<Integer, Integer, ArrayList<String>> boardState = createEmptyTable(rowNum, colNum);
+
+        // Fill board state with agents' string representations according to query result.
+        for (Entity entity : boardStateEntities) {
+            int row = ((Long) entity.getProperty("rowNum")).intValue();
+            int col = ((Long) entity.getProperty("colNum")).intValue();
+            String type = (String) entity.getProperty("agentType");
+            Long id = (Long) entity.getProperty("agentId");
+            boardState.get(row, col).add(type + id);
+        }
+
+        return gson.toJson(boardState);
+    }
+
+    private ArrayTable<Integer, Integer, ArrayList<String>> createEmptyTable(int rowNum, int colNum) {
+        ArrayTable<Integer, Integer, ArrayList<String>> table;
+        table = ArrayTable.create(IntStream.range(0, rowNum).boxed().collect(Collectors.toList()),
+                IntStream.range(0, colNum).boxed().collect(Collectors.toList()));
+
+        for (int row = 0; row < rowNum; row++) {
+            for (int col = 0; col < colNum; col++) {
+                table.set(row, col, new ArrayList<>());
+            }
+        }
+
+        return table;
+    }
+
+    private ArrayTable<Integer, Integer, ArrayList<String>> createEmptyTable(String simulationId) {
+        Entity simulationEntity = retrieveSimulationEntity(simulationId);
+        int rowNum = ((Long) simulationEntity.getProperty("rowNum")).intValue();
+        int colNum = ((Long) simulationEntity.getProperty("colNum")).intValue();
+        return createEmptyTable(rowNum, colNum);
+    }
+
+    private Entity retrieveSimulationEntity(String simulationId) {
+        Key simulationKey = KeyFactory.stringToKey(simulationId);
+        Query.Filter filterBySimulationId =
+                new Query.FilterPredicate(Entity.KEY_RESERVED_PROPERTY, Query.FilterOperator.EQUAL, simulationKey);
+        Query q = new Query("Simulation").setFilter(filterBySimulationId);
+        PreparedQuery pq = datastore.prepare(q);
+        return pq.asSingleEntity();
+    }
+
+    private boolean isRoundExistsInSimulation(String simulationId, int round) {
+        Entity simulationEntity = retrieveSimulationEntity(simulationId);
+        int maxSimulationRound = ((Long) simulationEntity.getProperty("roundsNum")).intValue();
+        return round < maxSimulationRound;
```

Comment:
I agree something needs to be changed, though I am not sure if we should change it here or at the Simulation's `run()` method (to be `currentRound < maxNumberOfRounds` as well).

It depends whether we consider the initial state (round zero) as a round, or not.

In my opinion it is more reasonable to count it as a round, otherwise we will have rounds 0 to maxRound recorded at out db - these are maxRound + 1 rounds instead of maxRound rounds (which is a bit weird, I think).

(In other words - I think we want the last round to be the maxRound - 1 round, thus we need to change the while loop condition of the simulation, and leave it as is here).

Reply:
I assume that considering the initial state as round 0 will be more comfortable, but it's your call. Just make sure it's consistent.

Reply:
Resolved at `omermadmon-simulation` branch :)